User guide fixes (#45)
kaitai-io/kaitai_struct_doc: user_guide.adoc

read, etc, etc, to name a few.

Kaitai Struct tries to isolate the developer from all these details
and allow them to focus on the things that matter: the data structure
itself, not particular ways to read or write it.

== Installation and invocation

Kaitai Struct has a somewhat diverse infrastructure around it. This
chapter will give an overview of the options available.

=== Web IDE

> If you're going to try Kaitai Struct for the first time, probably
> the Web IDE is the easiest way to get started. Just open

Reviewer: ```suggestion
If you're going to try Kaitai Struct for the first time,
the Web IDE is probably the easiest way to get started. Just open
```

https://ide.kaitai.io/[Kaitai Struct Web IDE] and you're
ready to go:

image::img/webide_png_example.png[caption="Figure 1: ", title="Kaitai Struct Web IDE (sample PNG file + png.ksy loaded)", alt="Kaitai Struct Web IDE (sample PNG file + png.ksy loaded)", width="1335", height="811"]

A list of Web IDE features is available on the https://github.com/kaitai-io/kaitai_struct_webide/wiki/Features[_kaitai_struct_webide_ GitHub wiki].

Note that there are two different versions of the Web IDE:

1. https://ide.kaitai.io/ — *stable* version, has the stable of the Kaitai Struct compiler (currently 0.9, released 2020-10-16)

> 2. https://ide.kaitai.io/devel/ — unstable development version, always has the *latest* compiler (the most recent 0.9-SNAPSHOT)

Reviewer: Yep, and please also update this:

```suggestion
2. https://ide.kaitai.io/devel/ — unstable development version, always has the *latest* compiler (the most recent 0.10-SNAPSHOT)
```

* `--verbose` — verbose output
* `--help` — display usage information and exit
* `--version` — display version information and exit

== Workflow overview

> The main idea
> of Kaitai Struct is that you create a description of a binary data

Reviewer: ```suggestion
The main idea of Kaitai Struct is that you create a description of a binary data
```

    contents: [0xca, 0xfe, 0xba, 0xbe]
----

This reads the first 4 bytes and compares them to the 4 bytes `CA FE BA BE`. If
there is any mismatch (or less than 4 bytes are read),

> that'll throw an exception and stop parsing at an early stage, before any

Reviewer: Would this be a bit better?

```suggestion
it throws an exception and stops parsing at an early stage, before any
```

some integer to designate length of the string, and store only

> designated number of bytes in the stream. Unfortunately, this yields
> variable-length structure, and it's impossible to describe such thing

Reviewer: ```suggestion
designated number of bytes in the stream. Unfortunately, this yields
a variable-length structure, and it's impossible to describe such thing
```

Another popular way to avoid allocating huge fixed-size buffers is to
use some sort of trailing delimiter. The most well-known example of
this is probably the null-terminated string which became a standard
string representation in C:

....
61 62 63 00
....

These 4 bytes actually represent the 3-character string "abc", plus one extra
trailing byte "0" (AKA null) which serves as a delimiter or

> terminator. By agreement, C strings cannot include a 0 byte: every time

Reviewer: ```suggestion
terminator. By agreement, C strings cannot include a byte 0: every time
```

Author: Not sure about this. I have a feeling (although haven't web searched to check), that people often refer to the "zero byte" in the context of string terminators.

Reviewer: > Not sure about this. I have a feeling (although haven't web searched to check), that people often refer to the "zero byte" in the context of string terminators.

Yeah, you're probably right, this was just my gut feeling, as I have never seen "0 byte" where 0 is written as a numeral. I'd say it's quite unusual. Maybe "zero byte" would be better?

Author: Yep, I've changed to "zero byte".

Of course, you can use any other byte (for example, `0xa`, AKA
newline) as a terminator. This gives Kaitai Struct some limited
capabilities to parse certain text formats as well.

Reading "until the terminator byte is encountered" could be
dangerous. What if we never encounter that byte?

Another very widespread model is actually having *both* a fixed-sized
buffer for a string *and* a terminator. This is typically an artifact

> of serializing structures like this from C. For example, if one would take this example structure:

Reviewer: Keep the line width:

```suggestion
of serializing structures like this from C. For example, if one would
take this example structure:
```

Author: Thanks for spotting this. I have resolved this in a different way.

[[enums]]
=== Enums (named integer constants)

The nature of binary format encoding dictates that in many cases
we'll be using some kind of integer constants to encode certain
entities. For example, an IP packet uses a 1-byte integer to

> encode protocol type for the payload: 6 would mean "TCP" (which gives us TCP/IP), 17 would mean "UDP" (which yields UDP/IP), and 1 means

Reviewer 2: I think there should be another "the" here:
```suggestion
entities. For example, an IP packet uses a 1-byte integer to
encode the protocol type for the payload: 6 would mean "TCP" (which gives us TCP/IP), 17 would mean "UDP" (which yields UDP/IP), and 1 means
```

----

Here we define another type named `str_with_len`, which we reference
just by doing `type: str_with_len`. The type itself is defined using the
`types:` key at the top level. That's a map, and inside it we can define as
many subtypes as we want. We define just one, and inside it we nest
the exact same syntax as we use for the type description on the top
level — i.e. the same `seq` designation.

> NOTE: There's no need for `meta` here, as the type name is derived from the

Reviewer: ```suggestion
NOTE: There's no need for `meta/id` here, as the type name is derived from the
```

=== Accessing attributes in other types

> Expression language (used, for example, in a `size` key) allows you to
> refer not only to attributes in the current type, but also in other types.

Reviewer 2: ```suggestion
The expression language (used, for example, in a `size` key) allows you to
refer not only to attributes in the current type, but also in other types.
```

----

> In this example, we again use expression language to specify a boolean

Reviewer 2: ```suggestion
In this example, we again use the expression language to specify a boolean
```

> * element repeated while some condition is satisfied (or until some
> condition won't become true)

Reviewer: To describe the `repeat-until` key, it should be:

```suggestion
* element repeated while some condition is not satisfied (or until some
condition becomes true)
```

Author: I am unclear on this...

Reviewer: > I am unclear on this...

I don't know how to formulate it clearly, but the attribute is repeated as long as the `repeat-until` condition is `false`, i.e. while the condition is not satisfied. An example can be found below in the User Guide (https://doc.kaitai.io/user_guide.html#_repeat_until_condition_is_met):

```ksy
seq:
  - id: numbers
    type: s4
    repeat: until
    repeat-until: _ == -1
```

And it usually translates to this code (JS):

```javascript
    do {
      var _ = this._io.readS4le();
      this.numbers.push(_);
    } while (!(_ == -1)); // `!` is a negation (NOT) operator
```

And another way how to express "while the condition is not satisfied (i.e. is equal to `false`)" is "until the condition **is** satisfied (i.e. evaluates to `true`)", because the word _until_ in English kind of means "while not ..." as I understand it.

So I think that _"until some condition won't become true"_ doesn't sound too English, because it can be rewritten to _"**while** some condition won't **not** become true"_, and we get a double negative here, which is not usual in English, and it apparently wasn't the intention anyway.

Author: My brain hurts, but I'll accept your suggestion anyway! It could always be improved another time if it needs it.

    repeat: eos
----

This yields an array of unsigned integers, each 4 bytes long, which
spans till the end of the stream. Note that if we've got a number of bytes left in the
stream that's not divisible by 4 (for example, 7), we'll end up reading
as much as possible, and the parsing procedure will throw an

> end-of-stream exception.

Reviewer 2: ```suggestion
as much as possible, and then the parsing procedure will throw an
end-of-stream exception.
```

    repeat-expr: width * height
----

This one specifies the `width` and `height` of the matrix first, then parses
as many `matrix` elements as needed to fill a `width` × `height` matrix
(although note that it won't be a true 2D matrix: it would still be just
a regular 1D array, and you'll need to convert (x, y) coordinates to

> addresses in that 1D array manually).

Reviewer: I'd prefer _index_ in the array rather than _address_ in the array:

```suggestion
a regular 1D array, and you'll need to convert (x, y) coordinates to
indices in that 1D array manually).
```

Memory _address_ is not exactly what we're looking for, because that is low level and can be implementation specific. Array index is a sufficiently high level (or _abstract_) and established term 🙂

----

In this format, instead of specifying just the identifier for every
numeric value, you specify a YAML map, which has an `id` key for

> an identifier, and allows other regular keys (like `doc` and `doc-ref`)

Reviewer 2: I think it would make more sense this way:

```suggestion
numeric value, you specify a YAML map, which has an `id` key for
the identifier, and allows other regular keys (like `doc` and `doc-ref`)
```

* `iso` specifies an ISO/IEC standard number, a reference to a standard
  accepted and published by https://www.iso.org/[ISO] (International
  Organization for Standardization). Typically these standards are not
  available for free (i.e. one has to pay to get a copy of a standard
  from ISO), and it's non-trivial to link to the ISO standards
  catalogue. However, ISO standards typically have clear designations
  like "ISO/IEC 15948:2004", so the value should cite everything
  except for "ISO/IEC", e.g. `15948:2004`.
* `justsolve` specifies an article name at
  http://fileformats.archiveteam.org/wiki/Main_Page["Just Solve the File Format
  Problem" wiki], a wiki that collects information on many file
  formats. A full link could be generated as
  `http://fileformats.archiveteam.org/wiki/` + this value.
* `loc` specifies an identifier in the
  https://www.loc.gov/preservation/digital/formats/fdd/browse_list.shtml[Digital
  Formats] database of the https://www.loc.gov/[US Library of Congress], a
  major effort to enumerate and document many file formats for digital
  preservation purposes. The value typically looks like `fddXXXXXX`, where
  `XXXXXX` is a 6-digit identifier.

> * `mime` specified s

Reviewer 2: Typo?

```suggestion
* `mime` specifies a
```

=== Processing: dealing with compressed, obfuscated and encrypted data

Some formats obscure the data fully or partially with techniques like
compression, obfuscation or encryption. In theses cases, incoming data

> should be pre-processed before actual parsing takess place, or we'll

Reviewer 2: ```suggestion
compression, obfuscation or encryption. In these cases, incoming data
should be pre-processed before actual parsing takes place, or we'll
```

use something more concise, such as `is_foo`.

*Byte arrays* are defined in the attribute syntax when you don't
specify anything as `type`. The size of a byte array is thus determined
using `size`, `size-eos` or `terminator`, one of which is mandatory in
this case. Byte array literals use typical array syntax like the one
used in Python, Ruby and JavaScript: i.e. `[1, 2, 3]`. There is a
little catch here: the same syntax is used for "true" arrays of
objects (see below), so if you try to do stuff like `[1, 1000, 5]`
(`1000` obviously won't fit in a byte), you won't get a byte array,
you'll get an array of integers instead.

*Strings* normally come from using `type: str` (or `type: strz`, which

> is the shortcut that implicitly adds `terminator: 0`).

Reviewer 2: ```suggestion
is a shortcut that implicitly adds `terminator: 0`).
```

* `a % b` — modulo; note that it's not a remainder: `-5 % 3` is `1`,
  not `-2`; the result is undefined for negative `b`.

NOTE: If both operands are integer, the result of an arithmetic operation is

> integer, otherwise it is a floating point number. For example, that

Reviewer 2: I think using "integer" as an adjective is grammatically correct, but in a programming context I almost always see it used as a noun. I would perhaps write it like this for clarity:

```suggestion
NOTE: If both operands are integers, the result of an arithmetic operation is
an integer, otherwise it is a floating point number. For example, that
```

to use is `obj.method(param1, param2, ...)`, which can be abbreviated
to `obj.method` if no parameters are required.

Note that when the `obj` in question is a user-defined type, you can
access all its attributes (both sequence and instances) using the same
`obj.attr_name` syntax. One can chain that to traverse a
chain of substructures: `obj.foo.bar.baz` (given that `obj` is a
user-defined type that has a `foo` field, which points to a user-defined
type that has a `bar` field, and so on).

There are a few pre-defined methods that form a kind of "standard

> library" for expression language.

Reviewer 2: ```suggestion
library" for the expression language.
```

|`to_i`
|Integer

> |Converts an enum into a corresponding integer representation

Reviewer 2: ```suggestion
|Converts an enum into the corresponding integer representation
```

(every enum value has exactly one corresponding integer value)

==== Switching over strings

> One can use type switching over other comparable values, not just integers. For example, one can switch over a string value. Note
> that the left side (key) of a `cases` map is a full-featured Kaitai Struct expression,

Reviewer: ```suggestion
One can use type switching over other comparable values,
not just integers. For example, one can switch over a string value. Note
that the left side (key) of a `cases` map is a full-featured Kaitai Struct expression,
```

|====

But sometimes you indeed want that dot, and it should be part of the

> output. This is possible, by specifying `include: true`

Reviewer: Maybe leave the comma out?

```suggestion
output. This is possible by specifying `include: true`
```

Yeah, this also makes sense with regard to this sentence a few lines below _(NB: no comma)_:

https://github.com/kaitai-io/kaitai_struct_doc/blob/f4ec2562de5bf4f002411e9d0b78b397b7c688a5/user_guide.adoc#L3059-L3060

[[eos-error]]
==== Ignoring errors in delimited structures

Delimited structures actually can be pretty scary: if we read until
the terminator is encountered, what will happen if we never find
one? In raw C, a typical answer is a "segmentation fault" occurring in
the operation that deals with such strings due to runaway reads past
the buffer, i.e. a very bad thing and a big security hole.

In Kaitai Struct, however, you can control this behavior. By default,
if you parse a structure awaiting a terminator, and that
terminator never happens, you'll get a clear error / exception similar
to "attempted to read past end of stream".

In some cases, you
really want to read until the end of the stream, and this should be
considered normal encoding, not an error. For these cases, you can

> specify `eos-error: false` (the default is true):

Reviewer: I'd go for

```suggestion
specify `eos-error: false` (the default is `true`):
```

... so that it has the same typeface as the value `false` a few words before.

----

and you want to use it in a file listing specification

> `filelist.ksy`, here's how to do that:

Reviewer: Good catch, now it makes a lot more sense! But I also suggest to replace _given that_ with _if_ at the beginning of the sentence:

> For example, ~~given that~~ **if** you have a `date.ksy` file that describes the date structure
> _[`date.ksy` contents]_
> and you want to use it in a file listing specification, here's how to do that:

Author: Done.

this.buf = _process__raw_buf.decode(this._raw_buf);
----

A typical implementation of a custom processor would look like this

> (again, example in Java; refer to language-specific notes for

Reviewer: A random thought (I **do not suggest** to address it in this PR at all): I remember that @GreyCat somewhere said it would be nice to show all supported target languages (i.e. C++, C#, Python, ...) for all code samples which show the generated code in one particular target language (mostly Java, currently), using tabs for switching among them.

Author: Sounds a good idea.

  this particular example, it just adds whatever we supplied as "key"
  to every byte

NOTE: `decode` can return a different number of bytes that it gets

> (which is perfectly normal, for example, with decompression).

Reviewer: I don't get this one (maybe it should be like this, but I'm not sure). Is there a rationale behind it, or your feeling?

Author: I've modified these lines again. Better now?

        if: has_right_child != 0
----

We've changed only three lines. We've enforced the parent of the node in the
`left_child` and `right_child` attributes to be passed as `_parent`,
not `this`. This, effectively, continues passing a reference to the original
node's parent, which is a `tree` type object, over and over in the whole
recursive structure. This way one can access the structure's root by just
using `_parent`. Naturally, we've done exactly that to get ourselves
`chunk_size` by just using `size: _parent.chunk_size`.

==== Omitting parent

> In some cases, you want some object to not have any parent

Reviewer: Shouldn't it be

```suggestion
In some cases, you want some object not to have any parent
```

?

Or maybe

> you want some object to ~~not~~ have ~~any~~ **no** parent at all.

Author: Thanks. I've gone for your second suggestion.

* `bool` to specify booleans
* `struct` to allow arbitrary KaitaiStruct-compatible user types
* `io` to specify KaitaiStream-compatible IO streams
* `any` to allow any type (if the target language supports that)

NOTE: Parametric types with mandatory parameters can't be read from

> files directly (as they obviously need parameter values supplied from

Reviewer: I would actually prefer the singular form, i.e. _from **a** file_ or _from **the** file_. "Reading from files" might sound like we want to read a single type from several files at once, which is not the case.

Author: Yep, agreed. Done.

....

In this example, we have N = 3 offsets, which specify N - 1 = 2
sections. Sizes of the sections must be derived as the start of the

> following section minus start of the current section. To do such

Reviewer 2: ```suggestion
following section minus the start of the current section. To do such
```
